xapian-letor/feature/colltfcolllenfeature.cc

```diff
@@ -41,33 +40,35 @@ CollTfCollLenFeature::get_values() const
 {
     LOGCALL(API, std::vector<double>, "CollTfCollLenFeature::get_values", NO_ARGS);
 
-    Query query = Feature::internal->feature_query;
-    map<string, long int> coll_tf = Feature::internal->collection_termfreq();
-    map<string, long int> coll_len = Feature::internal->collection_length();
-
     vector<double> values;
     double value = 0;
 
-    for (Xapian::TermIterator qt = query.get_terms_begin(); qt != query.get_terms_end(); ++qt) {
+    for (Xapian::TermIterator qt = feature_query.get_unique_terms_begin();
+	 qt != feature_query.get_terms_end(); ++qt) {
 	if ((*qt).substr(0, 1) == "S" || (*qt).substr(1, 1) == "S")
-	    value += log10(1 + ((double)coll_len["title"] / (double)(1 + coll_tf[*qt])));
+	    value += log10(1 + ((double)collection_length.find("title")->second
+				/ (double)(1 + collection_termfreq.find(*qt)->second)));
```

Comment:
If this `find()` doesn't find anything, then you dereference the end iterator which is undefined behaviour.

Now that should be "impossible" if I follow the code correctly, but that relies on a high level invariant which is not explicitly asserted (that all terms involved have an entry in `collection_termfreq`).  It would be all too easy for someone changing this code to stop that being the case - this is an example of what's sometimes called "code smell": https://en.wikipedia.org/wiki/Code_smell

My thought is that we should stop filling in 0 entries into these maps when the term isn't present in `FeatureList::Internal::compute_termfreq()` etc.  Then here do the `find()` just before the actual calculation and assume zero if the entry isn't found.  That avoids the potential for a crash, and saves some space in the maps.  Smaller maps also means faster lookup (because `std::map` lookup is O(log(N)) where N is the number of entries).

Same issue applies below too.